pkg/agent/nodeportlocal/portcache/port_table_windows_test.go

```diff
@@ -0,0 +1,164 @@
+			PodPort:  1002,
+			PodIP:    podIP,
+			Protocol: "udp",
+		},
+	}
+
+	mockPortRules.EXPECT().AddRule(nodePort1, podIP, 1001, "tcp")
+	mockPortRules.EXPECT().AddRule(nodePort1, podIP, 1001, "udp")
+	mockPortRules.EXPECT().AddRule(nodePort2, podIP, 1002, "udp")
+
+	syncedCh := make(chan struct{})
+	err := portTable.RestoreRules(allNPLPorts, syncedCh)
+	require.NoError(t, err)
+}
+
+func TestDeleteRule(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockPortRules := rulestesting.NewMockPodPortRules(mockCtrl)
+	mockPortOpener := portcachetesting.NewMockLocalPortOpener(mockCtrl)
+	portTable := newPortTable(mockPortRules, mockPortOpener)
+	npData := &NodePortData{
+		NodePort: startPort,
+		PodIP:    podIP,
+		PodPort:  1001,
+		Protocol: ProtocolSocketData{
+			Protocol: "tcp",
+		},
+	}
+
+	portTable.addPortTableCache(npData)
+	mockPortRules.EXPECT().DeleteRule(startPort, podIP, 1001, "tcp")
+	err := portTable.DeleteRule(podIP, 1001, "tcp")
+	require.NoError(t, err)
+}
+
+func TestDeleteRulesForPod(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockPortRules := rulestesting.NewMockPodPortRules(mockCtrl)
+	mockPortOpener := portcachetesting.NewMockLocalPortOpener(mockCtrl)
+	portTable := newPortTable(mockPortRules, mockPortOpener)
+
+	npData := []*NodePortData{
+		{
+			NodePort: startPort,
+			PodIP:    podIP,
+			PodPort:  1001,
+			Protocol: ProtocolSocketData{
+				Protocol: "tcp",
+			},
+		},
+		{
+			NodePort: startPort + 1,
+			PodIP:    podIP,
+			PodPort:  1002,
+			Protocol: ProtocolSocketData{
+				Protocol: "udp",
+			},
+		},
+	}
+
+	for _, data := range npData {
+		portTable.addPortTableCache(data)
+		mockPortRules.EXPECT().DeleteRule(data.NodePort, podIP, data.PodPort, data.Protocol.Protocol)
+	}
+
+	err := portTable.DeleteRulesForPod(podIP)
+	require.NoError(t, err)
+}
+
+func TestAddRule(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	mockPortRules := rulestesting.NewMockPodPortRules(mockCtrl)
+	mockPortOpener := portcachetesting.NewMockLocalPortOpener(mockCtrl)
+	portTable := newPortTable(mockPortRules, mockPortOpener)
+	podPort := 1001
+	tcs := []struct {
+		name        string
+		nodePort    int
+		protocol    string
+		expectedErr string
+	}{
+		{
+			name:     "valid entry for rule",
+			nodePort: startPort,
+			protocol: "udp",
+		},
+		{
+			name:        "duplicate rule",
+			nodePort:    startPort + 1,
+			protocol:    "udp",
+			expectedErr: "existing Windows Nodeport entry for",
+		},
+	}
+
+	for _, tc := range tcs {
+		t.Run(tc.name, func(t *testing.T) {
+			if tc.expectedErr == "" {
+				mockPortRules.EXPECT().AddRule(tc.nodePort, podIP, podPort, tc.protocol)
+			}
+			nodePort, err := portTable.AddRule(podIP, podPort, tc.protocol)
+			if tc.expectedErr != "" {
+				assert.ErrorContains(t, err, tc.expectedErr)
+			} else {
+				require.NoError(t, err)
+				assert.Equal(t, tc.nodePort, nodePort)
+			}
+		})
+	}
```

Comment:
table driven test's cases should be individual, so it's easy to add new cases and adjust existing ones with worrying about the others, and it makes it possible to run specific sub test only.
I assume in this test the 2nd subtest relies on the 1st one.
It should just do it sequentially to make the dependency clear:
```
mockCtrl := gomock.NewController(t)
defer mockCtrl.Finish()
mockPortRules := rulestesting.NewMockPodPortRules(mockCtrl)
mockPortOpener := portcachetesting.NewMockLocalPortOpener(mockCtrl)
portTable := newPortTable(mockPortRules, mockPortOpener)
podPort := 1001

// Adding the rule the first time should succeed.
mockPortRules.EXPECT().AddRule(tc.nodePort, podIP, podPort, tc.protocol)
gotNodePort, err := portTable.AddRule(podIP, podPort, tc.protocol)
require.NoError(t, err)
assert.Equal(t, startPort, gotNodePort)

// Add the same rule the second time should fail.
_, err := portTable.AddRule(podIP, podPort, tc.protocol)
assert.ErrorContains(t, err, "existing Windows Nodeport entry for")
```